### PR TITLE
refactor: Prepare for better guard typing from endo

### PR DESCRIPTION
**Pull request by erights**
Prepare for endos with https://github.com/endojs/endo/pull/1715 or https://github.com/endojs/endo/pull/1712 , while still being compat with agoric-sdk's current endo dependencies.

See also https://github.com/Agoric/agoric-sdk/pull/8212

### CODE REVIEW

File: packages/swingset-liveslots/src/virtualObjectManager.js [hunk 1]
```diff
@@ -954,6 +954,14 @@ export const makeVirtualObjectManager = (
...
         behavior,
         thisfulMethods,
+        // If annotating with the following at-ts-expect-error, then
+        // linting packages/telemetry complains of an unused
+        // at-ts-expect-error. But if it is removed, then linting
+        // here (swingset-liveslots) complains.
+        // at-ts-expect-error TS does not know it can only be a kit or undefined
+        //
+        // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error
+        // @ts-ignore
```
Comment (turadg) [inline packages/swingset-liveslots/src/virtualObjectManager.js]: I bet we can get these to agree but until then:
```suggestion
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error -- https://github.com/Agoric/agoric-sdk/issues/4620
    // @ts-ignore TS does not know it can only be a kit or undefined
```

File: packages/swingset-liveslots/src/virtualObjectManager.js [hunk 2]
```diff
@@ -962,6 +970,14 @@ export const makeVirtualObjectManager = (
...
```
Comment (turadg) [inline packages/swingset-liveslots/src/virtualObjectManager.js]:
```suggestion
    // eslint-disable-next-line @typescript-eslint/prefer-ts-expect-error -- https://github.com/Agoric/agoric-sdk/issues/4620
    // @ts-ignore TS does not know it can only be a kit or undefined
```

File: packages/swingset-liveslots/src/vatDataTypes.d.ts [hunk 3]
```diff
@@ -105,7 +104,9 @@ export type DefineKindOptions<C> = {
    * `prepareExoClass` use this internally to protect their raw class methods
    * using the provided interface.
    */
-  interfaceGuard?: InterfaceGuard<unknown>;
+  interfaceGuard?:
+    | InterfaceGuard<unknown>
+    | Record<string, InterfaceGuard<unknown>>;
```
Comment (turadg) [inline packages/swingset-liveslots/src/vatDataTypes.d.ts]: Can it really be either value or is this being loose with the cases?

IIUC it's a single `InterfaceGuard` unless it's a part of an Exo "kit" in which cases it's the Record. 
https://github.com/Agoric/agoric-sdk/blob/b9b86253042124e54a30c118f1eb39aa73f41a5a/packages/vat-data/src/exo-utils.js#L195-L200

WDYT of making that type more explicit? e.g. `interfaceGuard?: InterfaceGuard` and `interfaceGuards?: Record<string, InterfaceGuard<unknown>>`. Then in https://github.com/Agoric/agoric-sdk/blob/b9b86253042124e54a30c118f1eb39aa73f41a5a/packages/swingset-liveslots/src/virtualObjectManager.js#L951-L968 it `interfaceGuards` would be used in the multifaceted case.

Without this, the `defineKind` options are misleading to the developer.

File: packages/store/src/types.js [hunk 4]
```diff
@@ -1,96 +1,19 @@
...
-// Placeholder redundant types, to be imported from `@endo/patterns` instead.
-
-// TODO placeholder. Figure out how to import from `@endo/patterns` instead
-/** @typedef {Passable} Key */
-
-// TODO placeholder. Figure out how to import from `@endo/patterns` instead
-/** @typedef {Passable} Pattern */
-
-// TODO placeholder. Figure out how to import from `@endo/patterns` instead
+/** @typedef {import('@endo/pass-style').Passable} Passable */
```
Comment (turadg) [inline packages/store/src/types.js]: note that this is currently `any`. IMO while there the documentation value of naming it, it misleads the user into thinking they're getting some type annotation when in fact it's erasing any type information by propagating `any`.

Out of scope for this PR though. https://github.com/endojs/endo/issues/1488

Comment (erights) [reply packages/store/src/types.js]: Good point. Added a comment citing this issue.

File: packages/swingset-liveslots/src/vatDataTypes.d.ts (hunk 3 above)
Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: A nice improvement, done! Though rather than using a plural, I named the other one `interfaceGuardKit`.

Btw, I do hope to deprecate and eventually remove entirely the kind level of abstraction, leaving only the exo level. But not in this PR ;)

Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: Since this options name split is a dynamic change, please double check me on the following reasoning:

There is no upgrade issue with this dynamic change because we nowhere store a DefineKindOptions to virtual or durable storage. They are only used transiently during the construction of a kind class or a kind class kit.

There is no inter-repo version compat hazard because DefineKindOptions are only used internally within the agoric-sdk repo. Though they are used inter-package within that repo: between vat-data, swingset-liveslots, and zone. The zone package will be moving to endo, but has not yet, so we got lucky on that one.

Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: In the DefineKitOptions, as you suggest above, I also removed the `<undefined>` from the `interfaceGuard?:` and `interfaceGuardKit?:` types, since the `InterfaceGuard` type parameter already defaults to a decent type. Doing so made some type errors go away.

File: packages/swingset-liveslots/src/virtualObjectManager.js (hunk 1 above)
Comment (erights) [reply packages/swingset-liveslots/src/virtualObjectManager.js]: With your previous suggestion, these type errors disappeared. Good riddance!

File: packages/swingset-liveslots/src/virtualObjectManager.js (hunk 2 above)
Comment (erights) [reply packages/swingset-liveslots/src/virtualObjectManager.js]: With your previous suggestion, these type errors disappeared. Good riddance!

File: packages/store/src/types.js [hunk 5]
```diff
@@ -1,118 +1,44 @@
...
-
-// /////////////////////////////////////////////////////////////////////////////
-
 /**
  * @typedef {object} StoreOptions Of the dimensions on which KeyedStores can
  *   differ, we only represent a few of them as standard options. A given store
  *   maker should document which options it supports, as well as its positions
  *   on dimensions for which it does not support options.
- * @property {boolean} [longLived=true] Which way to optimize a weak store. True
+ * @property {boolean} [longLived] Which way to optimize a weak store. True
```
Comment (erights) [inline packages/store/src/types.js:26]: IIUC, we're also trying to remove default parameter values in these doc-comment positions. Don't we have a lint rule for that? (It did not trigger here)

So, this is a drive-by as long as I was in this file. Please let me know if I misunderstand. Thanks.

File: packages/swingset-liveslots/src/vatDataTypes.d.ts [hunk 6]
```diff
@@ -97,15 +96,34 @@ export type DefineKindOptions<C> = {
...
+   *
    * If an `interfaceGuard` is provided, then the raw methods passed alongside
    * it are wrapped by a function that first checks that this method's guard
    * pattern is satisfied before calling the raw method.
    *
-   * In `defineDurableKind` and its siblings, this defaults to off.
-   * `prepareExoClass` use this internally to protect their raw class methods
-   * using the provided interface.
+   * In `defineDurableKind` and its siblings, this defaults to `undefined`.
+   * Exo classes use this internally to protect their raw class methods
```
Comment (erights) [inline packages/swingset-liveslots/src/vatDataTypes.d.ts:111]: wrong. hold on while I fix this and re-push

File: packages/swingset-liveslots/src/vatDataTypes.d.ts [hunk 7]
```diff
@@ -97,15 +96,34 @@ export type DefineKindOptions<C> = {
...
+   * Intended for internal use only.
+   * Only applicable if this is a class kit kind. A class kind should use
+   * `interfaceGuard` instead.
+   *
+   * If an `interfaceGuardKit` is provided, then each member of the
+   * interfaceGuardKit is used to guard the corresponding facet of the
+   * class kit.
+   *
+   * In `defineDurableKindMulti` and its siblings, this defaults to `undefined`.
+   * Exo class kits use this internally to protect their facets, even if
```
Comment (erights) [inline packages/swingset-liveslots/src/vatDataTypes.d.ts]: wrong. hold on while I fix this and re-push

File: packages/swingset-liveslots/src/vatDataTypes.d.ts (hunk 6 above)
Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts:111]: Done

File: packages/swingset-liveslots/src/vatDataTypes.d.ts (hunk 7 above)
Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: Done

File: packages/swingset-liveslots/src/vatDataTypes.d.ts (hunk 3 above)
Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: Done, so resolving.

File: packages/store/src/types.js (hunk 5 above)
Comment (turadg) [reply packages/store/src/types.js:26]: I think it only triggers when annotating a function that has a default parameter. 

It's not coherent to me for an object typedef to have default values. It makes sense where `StoreOptions` is taken, but there's no stopping it from being used elsewhere.

File: packages/swingset-liveslots/src/vatDataTypes.d.ts [hunk 8]
```diff
@@ -97,15 +96,36 @@ export type DefineKindOptions<C> = {
...
+   * interfaceGuardKit is used to guard the corresponding facet of the
+   * class kit.
+   *
+   * In `defineDurableKindMulti` and its siblings, this defaults to `undefined`.
+   * Exo class kits use this internally to protect their facets.
+   * In absence, an exo is protected anyway, while a bare kind is
+   * not (detected by `!thisfulMethods`),
    */
-  interfaceGuard?: InterfaceGuard<unknown>;
+  interfaceGuardKit?: Record<string, InterfaceGuard>;
```
Comment (turadg) [inline packages/swingset-liveslots/src/vatDataTypes.d.ts]: good improvement.

this could be typed such that it takes _either_ `interfaceGuard` or `interfaceGuardKit` but I don't think it's worth it for this internal use.

Comment (turadg) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: incidentally, I find "Kit" as shorthand for "multifaceted" to be loose. I wonder if we can come up with a good word that is different from "kit" as used in "issuerKit" (foo and related).

File: packages/vat-data/src/exo-utils.js [hunk 9]
```diff
@@ -106,7 +106,7 @@ export const makeExoUtils = VatData => {
     });
   harden(defineVirtualExoClass);
 
-  // TODO interfaceGuard type https://github.com/Agoric/agoric-sdk/issues/6206
+  // TODO interfaceGuardKit type https://github.com/Agoric/agoric-sdk/issues/6206
```
Comment (turadg) [inline packages/vat-data/src/exo-utils.js]: done?

(same comment for other two)

**Review by turadg — APPROVED** [review]

File: packages/store/src/types.js (hunk 5 above)
Comment (erights) [reply packages/store/src/types.js:26]: Good. So done.

File: packages/swingset-liveslots/src/vatDataTypes.d.ts (hunk 8 above)
Comment (erights) [reply packages/swingset-liveslots/src/vatDataTypes.d.ts]: Would love to some suggested alternatives to "Kit". But we use it a lot so please short. In any case, because we use it a lot, it would need to be much better to be worth changing.

Resolving.

File: packages/vat-data/src/exo-utils.js (hunk 9 above)
Comment (erights) [reply packages/vat-data/src/exo-utils.js]: Good catch! Fixed a bunch of these. Thanks.
